### PR TITLE
feat(query): Optimize read system.columns data

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

If query `system.columns` with push down filter Expr FunctionCall eq

and the eq arg has column name 'table' like this:

```
where table='tb1' and table='tb2' and table='tb3' and table = 'tb4'
```

will only get these tables.

BTW, not use field database and Expr or as filter. Because there maybe have some bad case:

```
where database='db1' or table='db2_tb1'
where database='db1' and table='db1_tb1' or table='db2_tb1'
```

Closes #https://github.com/datafuselabs/databend/issues/11355

### CODE REVIEW

File: src/query/storages/system/src/table.rs [hunk 1]
```diff
@@ -233,12 +237,21 @@ impl<TTable: 'static + AsyncSystemTable> Table for AsyncOneBlockSystemTable<TTab
     fn read_data(
         &self,
         ctx: Arc<dyn TableContext>,
-        _: &DataSourcePlan,
+        plan: &DataSourcePlan,
         pipeline: &mut Pipeline,
     ) -> Result<()> {
         let inner_table = self.inner_table.clone();
+        let push_downs = plan.push_downs.clone();
+        println!("push_downs is {:?}", push_downs.clone());
```
Comment (b41sh) [inline src/query/storages/system/src/table.rs:245]: unused `println`

File: src/query/storages/system/src/columns_table.rs [hunk 2]
```diff
@@ -168,3 +219,28 @@ impl ColumnsTable {
...
+
+pub fn find_eq_table(expr: &Expr<String>, visitor: &mut impl FnMut(&str, &Scalar)) {
+    match expr {
+        Expr::Constant { .. } => {}
+        Expr::ColumnRef { .. } => {}
+        Expr::Cast { expr, .. } => {
+            find_eq_table(expr, visitor);
+        }
+        Expr::FunctionCall { function, args, .. } => {
+            if function.signature.name == "eq" {
```
Comment (b41sh) [inline src/query/storages/system/src/columns_table.rs:231]: Only check equal function may produce some unintended results, for example
```sql
where table='tb1' or table !='tb2';
```

Comment (TCeason) [reply src/query/storages/system/src/columns_table.rs:231]: Yes, this maybe cause some other bad case like :

```sql
where table='tb1' and table in ('tb2', 'tb3', 'tb4', 'tb5')
```

So I think filter only contains `eq` `and` `or` do this optimize.

File: src/query/storages/system/src/columns_table.rs [hunk 3]
```diff
@@ -134,17 +143,74 @@ impl ColumnsTable {
...
+                let expr = filter.as_expr(&BUILTIN_FUNCTIONS);
+                find_eq_db_table(&expr, &mut |col_name, scalar| {
+                    if col_name == "database" {
+                        if let Scalar::String(s) = scalar {
+                            if let Ok(database) = String::from_utf8(s.clone()) {
+                                databases.push(database);
+                            }
+                        }
+                    }
+                    if col_name == "table" {
```
Comment (b41sh) [inline src/query/storages/system/src/columns_table.rs:164]: `else if`

File: src/query/storages/system/src/columns_table.rs [hunk 4]
```diff
@@ -134,17 +143,74 @@ impl ColumnsTable {
...
+            } else {
+                let mut res = Vec::new();
+                for table in &tables {
+                    match catalog.get_table(tenant.as_str(), &database, table).await {
+                        Ok(table) => {
+                            res.push(table);
+                        }
+                        Err(_) => {
+                            debug!(
+                                "Can not get table {:?}.{:?}.{:?}",
```
[the remark this answers is not in the document]
Comment (TCeason) [reply src/query/storages/system/src/columns_table.rs:204]: That's ok, if table not exists we need ignore this err and generate the whole data. The return maybe empty.

**Review by b41sh — APPROVED** [review]

**Pull request merged**